File: test/MetaTest.java
// $Header$

// For j2meunit see http://j2meunit.sourceforge.net/
import j2meunit.framework.Test;
import j2meunit.framework.TestSuite;

// Auxiliary files
// MetaTest.lua - Lua source
// MetaTest.luc - MetaTest.lua compiled.


/**
 * J2MEUnit tests for Jili's metamethods.  DO NOT SUBCLASS.
 * public access granted only because j2meunit makes it necessary.
 */
public class MetaTest extends JiliTestCase
{
  /** void constructor, necessary for running using
   * <code>java j2meunit.textui.TestRunner MetaTest</code>
   */
  public MetaTest() { }

  /** Clones constructor from superclass.  */
  private MetaTest(String name)
  {
    super(name);
  }

  /**
   * Opens the base library into a fresh Lua state, calls a global
   * function, and returns the Lua state.
   * @param name  name of function to call.
   * @param n     number of results expected from function.
   */
  private Lua luaGlobal(String name, int n)
  {
    // :todo: push into superclass
    Lua L = new Lua();
    BaseLib.open(L);
    loadFile(L, "MetaTest");
    L.call(0, 0);
    System.out.println(name);
    L.push(L.getGlobal(name));
    L.call(0, n);
    return L;
  }

  /**
   * Calls a global lua function and checks that <var>n</var> results
   * are all true.
   */
  private void nTrue(String name, int n)
  {
    Lua L = luaGlobal(name, n);
    for (int i=1; i<=n; ++i)
    {
      assertTrue("Result " + i + " is true",
	  L.valueOfBoolean(true).equals(L.value(i)));
    }
  }

  public void testmetaindex0()
  {
    nTrue("testmetaindex0", 4);
  }

  public void testmetaindex1()
  {
    nTrue("testmetaindex1", 4);
  }

  public void testmetaindex2()
  {
    nTrue("testmetaindex2", 2);
  }

  public void testmetanewindex0()
  {
    nTrue("testmetanewindex0", 5);
  }

  public void testmetanewindex1()
  {
    nTrue("testmetanewindex1", 4);
  }

  public void testmetanewindex2()
  {
    nTrue("testmetanewindex2", 3);
  }

  public void testmetacall()
  {
    nTrue("testmetacall", 3);
  }

  public void testmetalt()
  {
    nTrue("testmetalt", 6);
  }

  public void testmetale()
  {
    nTrue("testmetale", 6);
  }

  public void testmetaadd()
  {
    nTrue("testmetaadd", 2);
  }

  public void testmetasub()
  {
    nTrue("testmetasub", 2);
  }

  public void testmetamul()
  {
    nTrue("testmetamul", 2);
  }

  public void testmetadiv()
  {
    nTrue("testmetadiv", 2);
  }

  public void testmetamod()
  {
    nTrue("testmetamod", 2);
  }

  public void testmetapow()
  {
    nTrue("testmetapow", 2);
  }

  public void testmetaconcat()
  {
    nTrue("testmetaconcat", 2);
  }

  public void testmetaunm()
  {
    nTrue("testmetaunm", 2);
  }

  public void testmetaeq()
  {
    nTrue("testmetaeq", 3);
  }

  public void testmetalen()
  {
    // :todo: implement me
  }

  public Test suite()
  {
    TestSuite suite = new TestSuite();

    suite.addTest(new MetaTest("testmetaindex0")
      {
        public void runTest() { testmetaindex0(); }
      });
    suite.addTest(new MetaTest("testmetaindex1")
      {
        public void runTest() { testmetaindex1(); }
      });
    suite.addTest(new MetaTest("testmetaindex2")
      {
        public void runTest() { testmetaindex2(); }
      });
    suite.addTest(new MetaTest("testmetanewindex0")
      {
        public void runTest() { testmetanewindex0(); }
      });
    suite.addTest(new MetaTest("testmetanewindex1")
      {
        public void runTest() { testmetanewindex1(); }
      });
    suite.addTest(new MetaTest("testmetanewindex2")
      {
        public void runTest() { testmetanewindex2(); }
      });
    suite.addTest(new MetaTest("testmetacall")
      {
        public void runTest() { testmetacall(); }
      });
    suite.addTest(new MetaTest("testmetalt")
      {
        public void runTest() { testmetalt(); }
      });
    suite.addTest(new MetaTest("testmetale")
      {
        public void runTest() { testmetale(); }
      });
    /* :todo:
    suite.addTest(new MetaTest("testmetaadd")
      {
        public void runTest() { testmetaadd(); }
      });
    suite.addTest(new MetaTest("testmetasub")
      {
        public void runTest() { testmetasub(); }
      });
    suite.addTest(new MetaTest("testmetamul")
      {
        public void runTest() { testmetamul(); }
      });
    suite.addTest(new MetaTest("testmetadiv")
      {
        public void runTest() { testmetadiv(); }
      });
    suite.addTest(new MetaTest("testmetamod")
      {
        public void runTest() { testmetamod(); }
      });
    suite.addTest(new MetaTest("testmetapow")
      {
        public void runTest() { testmetapow(); }
      });
    suite.addTest(new MetaTest("testmetaconcat")
      {
        public void runTest() { testmetaconcat(); }
      });
    suite.addTest(new MetaTest("testmetaunm")
      {
        public void runTest() { testmetaunm(); }
      });
    suite.addTest(new MetaTest("testmetaeq")
      {
        public void runTest() { testmetaeq(); }
      });
    suite.addTest(new MetaTest("testmetalen")
      {
        public void runTest() { testmetalen(); }
      });
    */
    return suite;
  }
}

File: test/MetaTest.lua
-- $Header$
-- Auxiliary file of Lua source for MetaTest.java
-- When you change this file, please compile it into MetaTest.luc and
-- submit both.

-- test __index for plain table case
function testmetaindex0()
  local t = {}
  local mt = {__index={foo='bar'}}
  setmetatable(t, mt)
  local a, b = t.a == nil, t.foo == 'bar'
  t.foo = 7
  local c = t.foo == 7
  t.foo = nil
  local d = t.foo == 'bar'
  return a,b,c,d
end
-- test __index for function case
function testmetaindex1()
  local t = {}
  local mt = {__index = function(t, k) return k*k end}
  setmetatable(t, mt)
  local a, b = t[3] == 9, t[7] == 49
  t[3] = 7
  local c = t[3] == 7
  t[3] = nil
  local d = t[3] == 9
  return a,b,c,d
end
-- Returns a function that when applied to a table returns new a new
-- table who (new) metatable has a field set to its original table
-- argument.  The field is variable and can be specified.
function newmeta(field)
  return function(x)
    local mt = {[field]=x}
    local t = {}
    setmetatable(t, mt)
    return t
  end
end
    
-- returns a new table whose (new) metatable has an __index value set to
-- argument
newmetaindex = newmeta('__index')
-- returns a new table whose (new) metatable has a __newindex values set
-- argument
newmetanewindex = newmeta('__newindex')
-- applies f to x n times, so returns f(f(f(...f(x)...)))
function niter(f, x, n)
  while n > 0 do
    x = f(x)
    n = n-1
  end
  return x
end
-- test __index for nested table case
function testmetaindex2()
  local deepindex = { foo='bar' }
  local t = niter(newmetaindex, deepindex, 20)
  local mt = getmetatable(t)
  local a = mt ~= getmetatable(mt.__index)
  local b = t.foo == 'bar'
  return a,b
end
-- test __newindex for table case
function testmetanewindex0()
  local t = {bar='spong'}
  local nit = {}
  local mt = { __newindex = nit }
  setmetatable(t, mt)
  t.foo = 'bar'
  t.bar = 'foo'
  local a,b = nit.foo == 'bar', t.foo == nil
  local c,d = nit.bar == nil, t.bar == 'foo'
  mt.__newindex = nil
  t.foo = 'spong'
  local e = t.foo == 'spong'
  return a,b,c,d,e
end
-- test __newindex for function case
function testmetanewindex1()
  local t = {}
  local f
  local update = {}
  do
    local i = 1
    f = function(t, k, v) update[i] = { [k] = v }; i=i+1; end
  end

  local mt = { __newindex = f }
  setmetatable(t, mt)
  t.foo = 'bar'
  t[3] = 7
  return t.foo == nil, t[3] == nil,
      update[1].foo == 'bar', update[2][3] == 7
end
-- test __newindex for nested table case
function testmetanewindex2()
  local deepnewindex = {}
  local t = niter(newmetanewindex, deepnewindex, 20)
  local mt = getmetatable(t)
  local a = mt ~= getmetatable(mt.__newindex)
  t.foo = 'bar'
  return t.foo == nil, deepnewindex.foo == 'bar', a
end
-- test __call
function testmetacall()
  local foocalled, beforef, afterf
  function around(f, g)
    local t = {}
    local mt = { __call = function()return g(f)end }
    setmetatable(t, mt)
    return t
  end
  function foo()foocalled = true end
  function a(f)beforef = true f() afterf = true end
  around(foo, a)()
  return foocalled, beforef, afterf
end
-- test __lt
function testmetalt()
  do
    local mt = { __lt = function(t1, t2) return #t1 < #t2 end }
    function addmt(t)
      setmetatable(t, mt)
      return t
    end
  end

  local a = addmt{}
  local b = addmt{'foo'}
  local c = addmt{'bar'}
  local d = addmt{a, b, c}

  local t = { a < b, b < a, b < c, c < b, c < d, d < c }  
  t[2] = not t[2]
  t[3] = not t[3]
  t[4] = not t[4]
  t[6] = not t[6]
  return unpack(t)
end
